src/lib.rs

```diff
@@ -410,54 +159,27 @@ impl SimpleAccumulator {
     /// In the third case, we replace the oldest element by the new element (FIFO order). All stats are updated online
     /// and the number of elements remains equal to the buffer capacity.
     ///
-    pub fn push<T: ToPrimitive>(&mut self, value: T) {
-        let y = T::to_f64(&value).unwrap();
-
+    pub fn push(&mut self, y: F) {
         // Running stats, Number of elements seen is incremented irrespective of buffer properties
         // Calculation is online following Knuth's algorithm
         self.total += 1;
 
-        let delta = y - self.mean;
-        let delta_n = delta / (self.total as f64);
-        self.mean += delta_n;
-
-        let term1 = delta * delta_n * (self.total as f64 - 1.0);
-        let stats1 = self.variance * (self.total as f64 - 2.0) + term1;
-
-        if self.total > 1 {
-            self.variance = stats1 / (self.total as f64 - 1.0);
-        }
+        self.min.update(y);
+        self.max.update(y);
+        self.sum.update(y);
+        self.mean.update(y);
+        self.median.update(y);
+        self.variance.update(y);
+        self.skewness.update(y);
+        self.kurtosis.update(y);
 
         // we just change the already held value and keep on rewriting it
         if self.fixed_capacity {
-            if self.is_empty() {
-                self.last_write_position = 0;
-            } else {
-                self.last_write_position = (self.last_write_position + 1) % self.capacity();
-            }
-
-            // Using vector push when ring buffer is not full
-            if self.len() < self.capacity() {
-                self.vec.push(y);
-            }
-            // Replacing first element by new element when ring buffer is full
-            else {
-                self.vec[self.last_write_position] = y;
-            }
-        }
-        // Using vector push in case fixed_capacity flag is 'false'
-        else {
-            self.vec.push(y);
-        }
-
-        if self.accumulate {
-            // Update stats for the buffer
-            if self.fixed_capacity {
-                self.calculate_all();
-            } else {
-                self.update_fields_increase(T::to_f64(&value).unwrap());
+            if self.data.len() == self.data.capacity() {
```

Comment:
If I input values `1, 2`, my mean would be `1.5` but now on fixed_capacity if  `3` is pushed, the new mean should have been `2.5` as my holding is `3, 2` won't this give us just `2` as mean?

Reply:
Add a test and check. The internal `watermill` library is doing the online calculation.

Reply:
https://docs.rs/watermill/0.1.1/src/watermill/mean.rs.html#46-49 

It doesn't seem to update it in a bounded manner, and I think you are popping the front and then adding something in back, it is 2 I/O ops compared to out initial one also followed by 2 more calculations meaning more time spend in I/O